services/app-api/utils/validation/completionStatus.ts

```diff
@@ -0,0 +1,167 @@
+import { CompletionData, AnyObject } from "../types/types";
+import { validateFieldData } from "./validation";
+
+export const isComplete = (completionStatus: CompletionData): Boolean => {
+  const flatten = (obj: AnyObject, out: AnyObject) => {
+    Object.keys(obj).forEach((key) => {
+      if (typeof obj[key] == "object") {
+        out = flatten(obj[key], out);
+      } else {
+        out[key] = obj[key];
+      }
+    });
+    return out;
+  };
+
+  const flattenedStatus = flatten(completionStatus, {});
+
+  for (const status in flattenedStatus) {
+    if (flattenedStatus[status] === false) {
+      return false;
+    }
+  }
+  return true;
+};
+
+// Entry point for calculating completion status
+export const calculateCompletionStatus = async (
+  fieldData: AnyObject,
+  formTemplate: AnyObject
+) => {
+  // Parent Dictionary for holding all route completion status
+
+  const validationJson = formTemplate.validationJson;
+
+  const areFieldsValid = async (
+    fieldsToBeValidated: Record<string, string>,
+    required: boolean
+  ) => {
+    let areAllFieldsValid = false;
+    try {
+      // all fields successfully validated if validatedFields is not undefined
+      areAllFieldsValid =
+        (await validateFieldData(
+          validationJson,
+          fieldsToBeValidated,
+          required
+        )) !== undefined;
+    } catch (err) {
+      // Silently ignore error, will result in false
+    }
+    return areAllFieldsValid;
+  };
+
+  const calculateFormCompletion = async (
+    nestedFormTemplate: AnyObject,
+    dataForObject: AnyObject = fieldData,
+    required: boolean = true
+  ) => {
+    // Build an object of k:v for fields to validate
+    let fieldsToBeValidated: Record<string, string> = {};
+    // Repeat fields can't be validated at same time, so holding their completion status here
+    let repeatersValid = true; //default to true in case of no repeat fields
+
+    // Iterate over all fields in form
+    for (var formField of nestedFormTemplate.fields || []) {
+      if (formField.repeat) {
+        // This is a repeated field, and must be handled differently
+        for (var repeatEntity of fieldData[formField.repeat]) {
+          // Iterate over each entity from the repeat section, build new value id, and validate it
+          repeatersValid &&= await areFieldsValid(
+            {
+              [formField.id]:
+                dataForObject[`${formField.id}_${repeatEntity.id}`],
+            },
+            required
+          );
+        }
+      } else {
+        // Key: Form Field ID, Value: Report Data for field
+        fieldsToBeValidated[formField.id] = dataForObject[formField.id];
+      }
+    }
+
+    // Validate all fields en masse, passing flag that uses required validation schema
+    return repeatersValid && areFieldsValid(fieldsToBeValidated, required);
+  };
+
+  const calculateEntityCompletion = async (
+    nestedFormTemplates: AnyObject[],
```

Comment:
Question: What is a nested form template? is it just a form? If so, [type here](https://github.com/Enterprise-CMCS/macpro-mdct-mcr/blob/e12110fa9ab9e2ad9f66d87e15b532816949c46e/services/ui-src/src/types/index.ts#L207-L213). Note: `formTemplate` is usually what the entire base form template (like mcpar.json or mlr.json) has been called.

Reply:
_Applies to other instances in file._

Reply:
it is the field templates, not the entire form. will rename to be clearer

Reply:
ehhh, so not just the field template. it is the drawerForm or modalDrawerForm object.

Reply:
Ah gotcha, that makes sense.